### PR TITLE
Implemet most of the database functions

### PR DESCRIPTION
**Pull request by kalbasit**
# implements
- notmuch_database_get_path
- notmuch_database_needs_upgrade
- notmuch_database_upgrade
- notmuch_database_begin_atomic and notmuch_database_end_atomic as one function with a callback to call
- notmuch_database_add_message
- notmuch_database_remove_message
- notmuch_database_find_message
- notmuch_database_find_message_by_filename
- notmuch_database_get_all_tags
- notmuch_database_compact
# does not implement
- notmuch_database_get_revision

### CODE REVIEW

File: db.go [hunk 1]
```diff
@@ -57,44 +60,153 @@ func Open(path string, mode DBMode) (*DB, error) {
...
+	cerr := C.notmuch_database_begin_atomic(db.cptr)
+	if err := statusErr(cerr); err != nil {
+		return err
+	}
+	callback(db)
+	cerr = C.notmuch_database_end_atomic(db.cptr)
+	if err := statusErr(cerr); err != nil {
+		return err
+	}
+	return nil
```
Comment (zenhack) [inline db.go]: Everything after the callback could be collapsed to:

```
return statusErr(C.notmuch_database_end_atomic(db.cptr))
```

File: db.go [hunk 2]
```diff
@@ -57,44 +60,153 @@ func Open(path string, mode DBMode) (*DB, error) {
...
 
 // LastStatus retrieves last status string for the notmuch database.
 func (db *DB) LastStatus() string {
-	return C.GoString(C.notmuch_database_status_string(db.toC()))
+	return C.GoString(C.notmuch_database_status_string(db.cptr))
 }
 
-func (db *DB) toC() *C.notmuch_database_t {
-	return (*C.notmuch_database_t)(db.cptr)
+// Path return the database path of the database.
```
Comment (zenhack) [inline db.go]: s/return/returns/

File: db.go [hunk 3]
```diff
@@ -57,44 +60,153 @@ func Open(path string, mode DBMode) (*DB, error) {
...
+
+// Upgrade upgrades the current database to the latest supported version. The
+// database must be opened with DBReadWrite.
+func (db *DB) Upgrade() error {
+	return statusErr(C.notmuch_database_upgrade(db.cptr, nil, nil))
+}
+
+// AddMessage adds a new message to the current database or associate an
+// additional filename with an existing message.
+func (db *DB) AddMessage(p string) (*Message, error) {
```
Comment (zenhack) [inline db.go]: Maybe change the parameter name to `filename`. It's more descriptive, and matches notmuch to boot.

File: db.go [hunk 4]
```diff
@@ -57,44 +60,153 @@ func Open(path string, mode DBMode) (*DB, error) {
...
+	cpath := C.CString(p)
+	defer C.free(unsafe.Pointer(cpath))
+
+	msg := &Message{}
+	cmsg := (**C.notmuch_message_t)(unsafe.Pointer(msg.cptr))
```
Comment (zenhack) [inline db.go]: This looks incorrect to me; I think you need to be taking the address of `msg.cptr`.

File: db.go [hunk 5]
```diff
@@ -57,44 +60,153 @@ func Open(path string, mode DBMode) (*DB, error) {
...
+	if msg.cptr == nil {
+		return nil, ErrNotFound
+	}
+	runtime.SetFinalizer(msg, func(m *Message) {
+		C.notmuch_message_destroy(m.cptr)
+	})
+	return msg, nil
+}
+
+// FindMessageByFilename finds a message with the given message_id.
```
Comment (zenhack) [inline db.go]: s/mesage_id/filename/

File: errors.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+// #include <stdlib.h>
+// #include <notmuch.h>
+import "C"
+import "errors"
+
+type status C.notmuch_status_t
+
+var (
+	// ErrOutOfMemory is returned when an Out of memory occured.
+	ErrOutOfMemory = statusErr(C.NOTMUCH_STATUS_OUT_OF_MEMORY)
```
Comment (zenhack) [inline errors.go:19]: Not sure we should expose this one; we should just be catching these and panicing.

**Comment by zenhack** [comment]
Re: removing checkOOM, the comments in `notmuch.h` state pretty clearly that this function will return NULL on OOM; we need to catch that. Not sure why you think it can be removed?

Other than the comments I made this looks pretty good to me. Good call on removing `toC()`; it was an artifact left over from a previous design which I'd discarded.

**Comment by kalbasit** [comment]
@zenhack addressed all of your comments. In regards to `checkOOM` I brought it back but I made it panic `ErrOutOfMemory` so the implementer has a chance to programmatically find out the cause of the panic (as opposed to a random string)

File: errors.go (hunk 6 above)
Comment (kalbasit) [inline errors.go:19]: I kept this one exposed and `checkOOM` now panic with it so we can compare it to the error received in recover.

**Comment by kalbasit** [comment]
@zenhack also I have no clue how to satisfy a `char**` argument in C. I'd like to implement `notmuch_database_get_revision` but I wasn't able to yesterday. Can you please write it up for me?

Signature for `notmuch_database_get_revision`:

```
/**
 * Return the committed database revision and UUID.
 *
 * The database revision number increases monotonically with each
 * commit to the database.  Hence, all messages and message changes
 * committed to the database (that is, visible to readers) have a last
 * modification revision <= the committed database revision.  Any
 * messages committed in the future will be assigned a modification
 * revision > the committed database revision.
 *
 * The UUID is a NUL-terminated opaque string that uniquely identifies
 * this database.  Two revision numbers are only comparable if they
 * have the same database UUID.
 */
unsigned long
notmuch_database_get_revision (notmuch_database_t *notmuch,
                const char **uuid);
```

**Comment by zenhack** [comment]
>    @zenhack also I have no clue how to satisfy a char*\* argument in C.
>    I'd like to implement notmuch_database_get_revision but I wasn't able
>    to yesterday. Can you please write it up for me?

Will do. I'll take a second look at the rest as soon as I can.

**Comment by kalbasit** [comment]
cool. I just implemented `notmuch_database_compact` and we have only revision left.

File: db_test.go [hunk 7]
```diff
@@ -66,3 +69,160 @@ func TestLastStatus(t *testing.T) {
...
+		t.Fatalf("error compacting %q: %s", dbPath, err)
+	}
+
+	db, err := Open(dbPath, DBReadOnly)
+	if err != nil {
+		t.Fatalf("Open(%q): unexpected error: %s", dbPath, err)
+	}
+	defer db.Close()
+
+	// same test as TestFindMessage to ensure the database integrity.
```
Comment (zenhack) [inline db_test.go]: Probably should factor the two out into a common function then.

**Comment by zenhack** [comment]
I hit a local test failure because NOTMUCH_STATUS_PATH_ERROR doesn't exist in the latest release; it's only on master (which is what travis has been testing). I changed the `.travis.yml` to test against the release branch, could you rebase off of master and fix this?

Other than that it looks good.

I'll note that there is more than just revision left; most of the tags stuff isn't there yet (but of course that needn't be part of this pr).

**Comment by kalbasit** [comment]
@zenhack IRC?

**Comment by kalbasit** [comment]
@zenhack tags stuff implemented in #7

File: db_test.go (hunk 7 above)
Comment (kalbasit) [inline db_test.go]: I usually don't do that in testing as I'd like the test to fully tell the story but I'm open to debate :)

**Pull request merged**